### PR TITLE
fix(USDT): allow BaseAdapter to support approving USDT

### PR DESCRIPTION
**Pull request by james-a-morris**
USDT on mainnet requires a spender's allowance to be zero prior to approving more value. 

This PR adds a zero'd approval if the token attempting to be approved is USDT.

### CODE REVIEW

**Comment by pxrl** [comment]
> USDT on mainnet requires a spender's allowance to be zero prior to approving more value.
> 
> This PR adds a zero'd approval if the token attempting to be approved is USDT.

Is this currently erroring somewhere?

File: src/common/Constants.ts [hunk 1]
```diff
@@ -295,3 +295,13 @@ export const chainIdsToCctpDomains: { [chainId: number]: number } = {
...
 };
+
+/**
+ * A mapping of chain IDs to tokens on that chain which need their allowance
+ * to first be zeroed before setting a new allowance. This is useful for
+ * tokens that have a non-standard approval process.
+ * @dev this is a generalization for USDT on Ethereum. Other tokens may be added
+ */
+export const TOKEN_APPROVALS_TO_FIRST_ZERO: Record<number, string[]> = {
+  1: ["0xdAC17F958D2ee523a2206206994597C13D831ec7"],
```
Comment (pxrl) [inline src/common/Constants.ts:306]: wdyt about referencing the existing constants we have?

```suggestion
     [CHAIN_IDs.MAINNET]: [TOKEN_SYMBOLS_MAP.USDT.addresses[CHAIN_IDs.MAINNET],
```

**Review by nicholaspai — APPROVED** [review]

**Comment by james-a-morris** [comment]
> > USDT on mainnet requires a spender's allowance to be zero prior to approving more value.
> > This PR adds a zero'd approval if the token attempting to be approved is USDT.
> 
> Is this currently erroring somewhere?

Yes - we're getting an error in the BaseAdapter when a new EOA needs to approve USDT

File: src/clients/bridges/BaseAdapter.ts [hunk 2]
```diff
@@ -161,22 +160,31 @@ export abstract class BaseAdapter {
...
     for (const { l1Token, targetContract } of tokensToApprove) {
-      const tx = await runTransaction(this.logger, l1Token, "approve", [targetContract, MAX_UINT_VAL]);
-      const receipt = await tx.wait();
       const { hubChainId } = this;
+      const txnRequests = [];
+      if (TOKEN_APPROVALS_TO_FIRST_ZERO[hubChainId]?.includes(l1Token.address)) {
+        txnRequests.push(runTransaction(this.logger, l1Token, "approve", [targetContract, bnZero]));
+      }
+      txnRequests.push(runTransaction(this.logger, l1Token, "approve", [targetContract, MAX_UINT_VAL]));
+      const txs = await Promise.all(txnRequests);
```
Comment (pxrl) [inline src/clients/bridges/BaseAdapter.ts:176]: Is there any chance of this causing nonce reuse issues? There'll be multiple points along the submission workflow where ethers blocks (i.e. when simulating the transaction to estimate the gasLimit, or potentially to query the transaction count for the EOA). Could this result in the same nonce being used for both transactions prior to submission?

(...I wonder if it's more robust to just submit them in sequence)

Comment (james-a-morris) [reply src/clients/bridges/BaseAdapter.ts]: This is fair - and there is also the case to consider that we need to guarantee that the zero-value approval happens first. I've made it sequential.

**Pull request merged**